Version 2.0.0-dev

caetera opened the pull request:
Preparation for release 1.5.0

Changes

- Code is fully migrated to .NET 8, thus, Mono dependency is obsolete
- Added an action for automatic building and testing on all code changes
- Added an action to produce self-contained and framework-based binaries for all major platforms on release
- Parquet format is updated to version 0.2
- Charge state array can be included in mzML and MGF output. (Close #189)

caetera committed:
Switching to Net 8.0

Known issues:
Parquet needs refactoring - new API
AWS S3 needs testing
Tests cannot be executed

caetera committed:
Tests updated to the latest NUnit

caetera committed:
Fix logging and mzML

* Log4net configuration moved to App.config
* Fixed unhandled exception during mzML precursor creation
* Explicit usings
* Properties (some build/debug information from VS) ignored

TODO Check all "possibly null reference" warnings

lazear committed:
feat: writer for mz_parquet v0.2 format

caetera committed:
Implement ion charges

caetera committed:
Convert AssemblyInfo to project properties

caetera committed:
Unified output path resolution

caetera committed:
Precursor logic streamlined

The same precursor selection logic as for other file formats
TBD: Move some of the service functions to the SpectrumWriter base class

caetera committed:
Unified mass spectrometry data

Use the same method for mass-spec data creation

caetera committed:
MS level filter and progress

caetera committed:
Move PrecursorTree to general SpectrumWriter

caetera committed:
Test for ParquetWriter

caetera committed:
Precursor scan fix for MS level restricted files

caetera committed:
Merge pull request #192 

mzParquet writer v0.2 full implementation

caetera committed:
Added build and test action

caetera committed:
Action to add artifacts on release

caetera committed:
Update Readme

caetera committed:
Tidying

caetera committed:
Refine error messages

caetera committed:
Zero-length mz data fix

caetera committed:
Precursor intensity in MGF

caetera committed:
Preserve file permissions when zipping artifacts

caetera committed:
Workaround for precursors mapped to missing scans

caetera committed:
Refactoring - move reading mz data to a separate function

caetera committed:
Updating Tests

caetera committed:
Refactor scan interval parsing and add NumberIterator

Replace custom scan interval parsing logic with a reusable NumberIterator class. Updates MainClass, QueryExecutor, and ProxiSpectrumReader to use NumberIterator for scan and MS level selection. MZArray.cs is merged into GeneralHelpers.cs.
Add unit tests for NumberIterator and scan interval handling.

caetera commented:
Features planned

* Support for scan slicing 
* Support for parent and neutral loss/gain scans

caetera committed:
First implementation of precursor ion spectrum

caetera committed:
Refactoring

caetera committed:
Support of precursor-ion scans to all formats

caetera committed:
Implementation of neutral loss/gain

timosachsenberg commented:
awesome :)

ypriverol commented:
@caetera do you think you can produce some examples of the parquet implementation to see how it looks, the idea is to have some parquet users evaluating the implementation.

caetera commented:
@ypriverol I certainly can convert a few files to parquet format. What kind of files do you want to test on? Are there any specific requirements?

ypriverol commented:
I think having a few examples from different instruments would be good to know how the format works. I can try to viz here, it is only to see if we need to extend the current implementation further.

ypriverol commented:
> I shared some files here - https://syddanskuni-my.sharepoint.com/:f:/g/personal/vgor_bmb_sdu_dk/EiXlHHyZY35LjvUCX_J9eJ0BJXMBbL7lqvQ2HkKqiyEzyA?e=Ikk9au converted with the latest build ([5bac912](https://github.com/CompOmics/ThermoRawFileParser/commit/5bac9120a2698c59d90c1af648a864862857b763))
> 
> File names are [Instrument]-[MS1Mode]-[MS2Mode]_[Other]
> 
> OTc - Orbitrap in centroided mode OTp - Orbitrap in profile mode ITC - Ion Trap in centroided mode ITp - Ion Trap in profile mode
> 
> I converted them to mzparquet with vendor centroiding enabled (default) and disabled.

Thanks I will take a look.

timosachsenberg commented:
@mobiusklein raises some good points, and the answer depends somewhat on the intended scope of mzParquet. As I understand it, mzParquet is primarily designed as a low-metadata format, for example to support quick visualization of spectra. If the scope were to be extended to use cases such as database searches, the mentioned columns could be valuable in enabling those applications. Ultimately, a more complete approach like mzpeak is needed to support all analyses - but I think this is not in the scope of that format.

lazear commented:
I view mzparquet similarly to mzML, MGF, MS2 - centroided input for a search engine, visualization tools, etc. IMO, these are 'necessary evil' intermediate file formats to sidestep proprietary vendor libraries for reading MS data, not intended to necessarily capture 100% of the metadata from the raw files - that information can just be extracted from the raw files if required. 

mzparquet reflects the need for a lighter-weight, higher performance mzML/MGF replacement. In contrast to the other formats, the data is readable with pandas/polars/distributed query engines/etc, making it easier to integrate data for both interactive data exploration use cases or repository-scale analyses. 

The nice thing about a parquet-defined format is that we could define the required minimum usable set of columns (m/z, intensity, rt, ion-mobility, etc) that cover 90% of use-cases. Additional columns could be added for mass analyzers or whatever other metadata is desired, and downstream applications that don't need those columns can simply skip reading them - it is essentially zero analysis-time cost (outside of storage space) to add new columns, something not possible with mzML, etc.

lazear commented:
AFAIK there are no public users of mzparquet (including sage). I have some internal tools that use it, and I can test it out. 

TRFP already had an obscure parquet based format that no one was using, so this was just a swap to another format no one is using yet :)

ypriverol commented:
@caetera, another thing. This new version of the tool is a big change in the architecture, no mono dependency anymore, new parquet, etc. Should we increase to 2.0.0 rather than continue in 1.X.X versions? What do you think @caetera @timosachsenberg @mobiusklein

caetera commented:
In the current implementation, mzparquet is well-suited for visualization, XIC extraction, etc, but if it is expected to be used also for search engines, adding columns like the type of activation, collision energy, supplementary activation type and energy, makes sense.

If we want someone to adopt it, I believe it will be necessary to present the schema somehow, i.e., required columns (names) and datatypes (`float32`/`float64` could be interchangeable, though, since the downstream readers will likely take care of that). For activation and other similar fields, I would suggest using a qualified CV term name (`collision-induced dissociation`, `electron transfer dissociation`, etc) to keep it lightweight.

caetera commented:
> @caetera, another thing. This new version of the tool is a big change in the architecture, no mono dependency anymore, new parquet, etc. Should we increase to 2.0.0 rather than continue in 1.X.X versions? What do you think @caetera @timosachsenberg @mobiusklein

Strictly speaking, since "the API" (dependencies and one of the output formats) changes, we should increment the major version; however, it does not change much "under the hood".

caetera commented:
I mean "the API" of TRFP, the API of Thermo libraries did not change between `dotnet` and `net-framework` version.

mobiusklein commented:
> If we want someone to adopt it, I believe it will be necessary to present the schema somehow, i.e., required columns (names) and datatypes (float32/float64 could be interchangeable, though, since the downstream readers will likely take care of that)

That's probably the case. For statically typed languages rather than query engine embedded in a higher level language, e.g. in C++ and Rust (where I have actually read/written things, but I believe also the case in C#), `Float32Array` and `Float64Array` are different types, so you have to effectively put a switch/if-else tree over operations of the same kind on slightly different width types.

caetera committed:
Update CV mappings for new instruments

Added new instrument CV mappings and detector assignments to follow PSI-MS 4.1.204

caetera commented:
Converted to Release 2.0.0-dev

Summary of changes

- Code is fully migrated to .NET 8, thus, Mono dependency is obsolete
- Added an action for automatic building and testing on all code changes
- Added an action to produce self-contained and framework-based binaries for all major platforms on release
- Parquet format is updated to version 0.2
- Charge state array can be included in mzML and MGF output. (Close #189)
- Precursor ion intensity in MGF output
- Support for precursor ion and neutral/loss gain scans (Close #197 and #198)
- Cleaner error messaging and improved error handling (Close #194 and #195)
- Introduce the latest instrument CV terms (PSI-MS 4.1.204)

caetera committed:
Convert to 2.0.0-dev
